misc/metrics/src/identify.rs

```diff
@@ -96,27 +112,136 @@ impl Metrics {
     }
 }
 
-impl super::Recorder<libp2p_identify::IdentifyEvent> for super::Metrics {
+impl super::Recorder<libp2p_identify::IdentifyEvent> for Metrics {
     fn record(&self, event: &libp2p_identify::IdentifyEvent) {
         match event {
             libp2p_identify::IdentifyEvent::Error { .. } => {
-                self.identify.error.inc();
+                self.error.inc();
             }
             libp2p_identify::IdentifyEvent::Pushed { .. } => {
-                self.identify.pushed.inc();
+                self.pushed.inc();
             }
-            libp2p_identify::IdentifyEvent::Received { info, .. } => {
-                self.identify.received.inc();
-                self.identify
-                    .received_info_protocols
+            libp2p_identify::IdentifyEvent::Received { peer_id, info, .. } => {
+                {
+                    let mut protocols: Vec<String> = info
+                        .protocols
+                        .iter()
+                        .filter(|p| {
+                            let allowed_protocols: &[&[u8]] = &[
+                                #[cfg(feature = "dcutr")]
+                                libp2p_dcutr::PROTOCOL_NAME,
+                                // #[cfg(feature = "gossipsub")]
+                                // #[cfg(not(target_os = "unknown"))]
+                                // TODO: Add Gossipsub protocol name
+                                libp2p_identify::PROTOCOL_NAME,
+                                libp2p_identify::PUSH_PROTOCOL_NAME,
+                                #[cfg(feature = "kad")]
+                                libp2p_kad::protocol::DEFAULT_PROTO_NAME,
+                                #[cfg(feature = "ping")]
+                                libp2p_ping::PROTOCOL_NAME,
+                                #[cfg(feature = "relay")]
+                                libp2p_relay::v2::STOP_PROTOCOL_NAME,
+                                #[cfg(feature = "relay")]
+                                libp2p_relay::v2::HOP_PROTOCOL_NAME,
+                            ];
+
+                            allowed_protocols.contains(&p.as_bytes())
+                        })
+                        .cloned()
+                        .collect();
+
+                    // Signal via an additional label value that one or more
+                    // protocols of the remote peer have not been recognized.
+                    if protocols.len() < info.protocols.len() {
+                        protocols.push("unrecognized".to_string());
+                    }
+
+                    protocols.sort_unstable();
+                    protocols.dedup();
```

Comment:
This will flatten all `unrecognized` into 1, regardless of how many unrecognized protocols there are.

Would it make sense to increment a separate counter for the number of unrecognized protocols?

Reply:
> Would it make sense to increment a separate counter for the number of unrecognized protocols?

We would then need to track the number of unrecognized protocols per peer. On each new identify message, we would have to subtract the previous amount and add the new amount to the counter. I am not saying this is impossible, just adds a bit of complexity which I am not sure is worth the benefit. In other words, is it relevant for a user to know how many, potentially duplicate, unrecognized protocols all connected peers offer?

I was hoping for the addition to the `# HELP` text to resolve any confusion:

> with "unrecognized" for each peer supporting one or more unrecognized protocols

Reply:
> > Would it make sense to increment a separate counter for the number of unrecognized protocols?
> 
> We would then need to track the number of unrecognized protocols per peer. On each new identify message, we would have to subtract the previous amount and add the new amount to the counter.

I think this would only apply if we would track it as a gauge?

A counter always increases anyway and you need to use functions like `rate` to see how it changes over time.

A spike in the unrecognized protocols _rate_ could e.g. hint at a spam attack by a peer.

I don't feel strongly about it though, happy to go either way :)

Reply:
> > > Would it make sense to increment a separate counter for the number of unrecognized protocols?
> > 
> > 
> > We would then need to track the number of unrecognized protocols per peer. On each new identify message, we would have to subtract the previous amount and add the new amount to the counter.
> 
> I think this would only apply if we would track it as a gauge?
> 
> A counter always increases anyway and you need to use functions like `rate` to see how it changes over time.

The rate on a counter would only tell how often you saw a unrecognized protocol, not how many unrecognized protocols connected peers offer. One could correlate this with the interval `libp2p-identify` requests a new identification, though that is (a) neither specified nor (b) solid with the push mechanism.

> A spike in the unrecognized protocols _rate_ could e.g. hint at a spam attack by a peer.

True, though I think we would want to prevent such attack before it happens. In addition, given that this attack would potentially bring down our Prometheus instance, who is going to alert us of the spam attack?

Reply:
> > > > Would it make sense to increment a separate counter for the number of unrecognized protocols?
> > > 
> > > 
> > > We would then need to track the number of unrecognized protocols per peer. On each new identify message, we would have to subtract the previous amount and add the new amount to the counter.
> > 
> > 
> > I think this would only apply if we would track it as a gauge?
> > A counter always increases anyway and you need to use functions like `rate` to see how it changes over time.
> 
> The rate on a counter would only tell how often you saw a unrecognized protocol, not how many unrecognized protocols connected peers offer. One could correlate this with the interval `libp2p-identify` requests a new identification, though that is (a) neither specified nor (b) solid with the push mechanism.

Assuming a stable set of peers, the rate of unrecognised protocols should be stable too? Identify pushes might throw a little spanner in here and there but overall, I'd expect it to be meaningful.

> > A spike in the unrecognized protocols _rate_ could e.g. hint at a spam attack by a peer.
> 
> True, though I think we would want to prevent such attack before it happens. In addition, given that this attack would potentially bring down our Prometheus instance, who is going to alert us of the spam attack?

We are preventing the attack by only increasing a counter instead of creating a label per protocol. Surely prometheus can't be brought down just because a counter increases at a massive rate?

All I am saying is that increasing the counter by the number of unrecognised protocols should allow us to observe the attempt of an attack assuming that a non-malicious node will have a stable number of unrecognised protocols whereas an attack would likely max out the size limit of the incoming message with as many protocols as possible and thus increase the counter by more than ordinary behaviour.

Reply:
> > > A spike in the unrecognized protocols _rate_ could e.g. hint at a spam attack by a peer.
> > 
> > 
> > True, though I think we would want to prevent such attack before it happens. In addition, given that this attack would potentially bring down our Prometheus instance, who is going to alert us of the spam attack?
> 
> We are preventing the attack by only increasing a counter instead of creating a label per protocol. Surely prometheus can't be brought down just because a counter increases at a massive rate?

Oh error in reasoning on my end here. You are right. Sorry about that @thomaseizinger.

> It is not massively useful outside of this I think so happy to go either way.

I will go without it.